hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/volume/OMVolumeDeleteRequest.java

```diff
@@ -0,0 +1,200 @@
+    } catch (IOException ex) {
+      LOG.error("Volume deletion failed for volume:{}", volume, ex);
+      omMetrics.incNumVolumeDeleteFails();
+      auditLog(auditLogger, buildAuditMessage(OMAction.DELETE_VOLUME,
+          buildVolumeAuditMap(volume), ex, userInfo));
+      return new OMVolumeDeleteResponse(null, null, null,
+          createErrorOMResponse(omResponse, ex));
+    } finally {
+      omMetadataManager.getLock().releaseVolumeLock(volume);
+    }
+
+    // We cannot acquire user lock holding volume lock, so released volume
+    // lock, and acquiring user and volume lock.
+
+    omMetadataManager.getLock().acquireUserLock(owner);
+    omMetadataManager.getLock().acquireVolumeLock(volume);
+
+    String dbUserKey = omMetadataManager.getUserKey(owner);
+    String dbVolumeKey = omMetadataManager.getVolumeKey(volume);
+
+    IOException exception = null;
+    OzoneManagerProtocolProtos.VolumeList newVolumeList = null;
+    try {
+      if (!omMetadataManager.isVolumeEmpty(volume)) {
+        LOG.debug("volume:{} is not empty", volume);
+        throw new OMException(OMException.ResultCodes.VOLUME_NOT_EMPTY);
+      }
+
+      newVolumeList = omMetadataManager.getUserTable().get(owner);
+
+      // delete the volume from the owner list
+      // as well as delete the volume entry
+      newVolumeList = delVolumeFromOwnerList(newVolumeList, volume, owner);
+
+      omMetadataManager.getUserTable().addCacheEntry(new CacheKey<>(dbUserKey),
+          new CacheValue<>(Optional.of(newVolumeList), transactionLogIndex));
+
+      omMetadataManager.getVolumeTable().addCacheEntry(
+          new CacheKey<>(dbVolumeKey), new CacheValue<>(Optional.absent(),
+              transactionLogIndex));
+
+    } catch (IOException ex) {
+      exception = ex;
+
+    } finally {
+      omMetadataManager.getLock().releaseVolumeLock(volume);
+      omMetadataManager.getLock().releaseUserLock(owner);
+    }
+
+    // Performing audit logging outside of the lock.
+    auditLog(auditLogger, buildAuditMessage(OMAction.DELETE_VOLUME,
+        buildVolumeAuditMap(volume), exception, userInfo));
+
+    // return response after releasing lock.
+    if (exception == null) {
+      LOG.debug("Volume deleted for user:{} volume:{}", owner, volume);
+      omMetrics.decNumVolumes();
+      omResponse.setDeleteVolumeResponse(
+          DeleteVolumeResponse.newBuilder().build());
+      return new OMVolumeDeleteResponse(volume, owner, newVolumeList,
+          omResponse.build());
+    } else {
+      LOG.error("Volume deletion failed for user:{} volume:{}",
+          owner, volume, exception);
+      omMetrics.incNumVolumeDeleteFails();
+      return new OMVolumeDeleteResponse(null, null, null,
+          createErrorOMResponse(omResponse, exception));
+    }
+
+  }
+
+  private OmVolumeArgs getVolumeInfo(OMMetadataManager omMetadataManager,
+      String volume) throws IOException {
+
+    omMetadataManager.getLock().acquireVolumeLock(volume);
```

Comment:
getVolumeInfo is only called in one place and with the lock held. I think it's okay to add a Javadoc that this method must be invoked with the volume lock held, and avoid reacquiring the lock once more.

Reply:
done.